include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+/*
+ * Copyright 2022 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OSSL_INTERNAL_EVENT_H
+# define OSSL_INTERNAL_EVENT_H
+# pragma once
+
+# include <stddef.h>
+# include <stdarg.h>
+# include <openssl/safestack.h>
+# include "crypto/sparse_array.h"
+
+/*-
+ * Types and structures
+ * --------------------
+ */
+
+/*
+ * An event is expected to be very short lived, and may be a static
+ * (re-usable) variable, local in a function or other structure, or
+ * dynamically allocated.
+ */
+struct ossl_event_st;
+typedef int ossl_event_destructor_fn(struct ossl_event_st *event);
+
+/*
+ * Generic callback function type.  This is used for subscribers and filters
+ * alike, as seen in the different functions where this type is used.
+ */
+typedef int ossl_event_callback_fn(struct ossl_event_st *event,
+                                   void *callback_data);
+
+struct ossl_event_st {
+    /*
+     * Mandatory data.
+     *
+     * These fields are considered immutable.  However, the contents that
+     * |ctx| points at may be mutable.
+     */
+
+    /* A numeric event identity, defined by the using functionality */
+    uint32_t type;
+    /* A context relevant for the using functionality */
+    void *ctx;
+
+    /*
+     * Optional data (may be NULL).
+     *
+     * These fields are considered mutable, in so far that an event handler
+     * may "steal" them, and replace them with NULL.
+     */
+
+    /* Identifying material */
+    const void *identifiers;
+    /* Data carried by the event */
+    size_t payload_size;
+    void *payload;
+
+    /*
+     * A destructor for |identifiers| and |payload|, if necessary.
+     *
+     * This field is considered immutable until the event is destroyed.
+     */
+    ossl_event_destructor_fn *destructor;
+
+    /*
+     * Flags, only relevant for the event functionality itself.
+     */
+    unsigned int flag_dynamic : 1;
```

Comment:
I think it would be cleaner to have a different destructor for an allocated event rather than a flag.

Reply:
The destructor that's passed is for destroying the contents, but not the event structure itself.  That makes it depend on what was called to create/populate an event, which I was thinking would be one less thing for the programmer that uses this to think about (i.e. hopefully less memory leaks)

Reply:
Hmmm, the caller explicitly creates the event structure but the event queue frees it?  Seems a bit asymmetric to me.  It also prevents the caller from reusing it.

I'm happy to live with this but it feels a bit uncomfortable.

Reply:
yeah ok, I can remake it the way you suggest

Reply:
The more I think about automatic free, the less I like it.  It prevents reusing the event record, forcing a new allocation.  It doesn't prevent a caller from needing to free the event -- what happens on error before the event is processed?